pkg/socketlb/cgroup.go

```diff
@@ -0,0 +1,205 @@
+// On upgrade, a link can be updated using link.Update(), which will atomically replace the
+// currently running bpf program.
+//
+// PROG_ATTACH (all kernel versions pre 5.7 that cilium supports): by definition the association
+// outlives userspace as the cgroup will hold a reference  to the attached program and detaching
+// must be done explicitly using PROG_DETACH.
+// This API is what cilium has been using prior to the 1.14 release and will continue to use if
+// bpf_link is not available.
+// On upgrade, cilium will continue to seamlessly replace old programs with the PROG_ATTACH API,
+// because updating it with a bpf_link could cause connectivity interruptions.
+
+package socketlb
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"golang.org/x/sys/unix"
+)
+
+var attachTypes = map[string]ebpf.AttachType{
+	Connect4:     ebpf.AttachCGroupInet4Connect,
+	SendMsg4:     ebpf.AttachCGroupUDP4Sendmsg,
+	RecvMsg4:     ebpf.AttachCGroupUDP4Recvmsg,
+	GetPeerName4: ebpf.AttachCgroupInet4GetPeername,
+	PostBind4:    ebpf.AttachCGroupInet4PostBind,
+	PreBind4:     ebpf.AttachCGroupInet4Bind,
+	Connect6:     ebpf.AttachCGroupInet6Connect,
+	SendMsg6:     ebpf.AttachCGroupUDP6Sendmsg,
+	RecvMsg6:     ebpf.AttachCGroupUDP6Recvmsg,
+	GetPeerName6: ebpf.AttachCgroupInet6GetPeername,
+	PostBind6:    ebpf.AttachCGroupInet6PostBind,
+	PreBind6:     ebpf.AttachCGroupInet6Bind,
+}
+
+// attachCgroup attaches a program from spec with the given name to cgroupRoot.
+// If the kernel supports it, the resulting bpf_link is pinned to pinPath.
+//
+// Upgrades from prior Cilium versions will continue to be handled by a PROG_ATTACH
+// to replace an old program attached to a cgroup.
+func attachCgroup(spec *ebpf.Collection, name, cgroupRoot, pinPath string) error {
+	prog := spec.Programs[name]
+	if prog == nil {
+		return fmt.Errorf("program %s not found in ELF", name)
+	}
+
+	// A previously created link is found and updated.
+	pin := filepath.Join(pinPath, name)
+	l, err := link.LoadPinnedLink(pin, &ebpf.LoadPinOptions{})
+	if err == nil {
+		if err := l.Update(prog); err != nil {
+			return fmt.Errorf("update link %s: %w", pinPath, err)
+		}
+
+		log.Debugf("updated link %s for program %s", pinPath, name)
+		return nil
+	}
+	log.Debugf("no pinned link at '%s'", pinPath)
+
+	cg, err := os.Open(cgroupRoot)
+	if err != nil {
+		return fmt.Errorf("open cgroup %s: %w", cgroupRoot, err)
+	}
+	defer cg.Close()
+
+	// no previous link is found, try to create one. This will only work on fresh installs.
+	l, err = link.AttachRawLink(link.RawLinkOptions{
+		Target:  int(cg.Fd()),
+		Program: prog,
+		Attach:  attachTypes[name],
+	})
+
+	switch {
+	// Hit a recoverable error that could be resolved by attaching the prog
+	// with PROG_ATTACH.
+	//
+	// Kernels before 5.7 don't support bpf_link.
+	// In that case link.AttachRawLink returns ErrNotSupported.
```

Comment:
Not sure if this was discussed before, but why do rely on being able to create a link or not as a way to determine whether the agent attaches cgroup bpf programs via link or `PROG_ATTACH`? Can we not lean on the bpftool probe approach we use everywhere else in the agent?

Reply:
> why do we rely on being able to create a link or not as a way to determine whether the agent attaches cgroup bpf programs via link or PROG_ATTACH?

Sure, we can discuss this, but this might not be the right medium, so bear with me. There's a concept in programming called 'ask for forgiveness, not permission'. Attempt to use the newest feature, and gradually fall back until something works. This has a few benefits:
- code simplicity and readability: fewer branches + more early returns = lower cognitive overhead.
- efficiency: in this case you mention a probe, but why probe if we can directly attempt the real thing instead? Fewer syscalls made and less time spent waiting on them to complete. The probe and the actual syscall convey mostly the same information.
- avoiding TOCTOU: not 100% applicable to the code here, but I see many cases of stat+open littered throughout the Cilium codebase. Stat is expensive, and an arbitrary amount of time can pass until open() is called, meaning the file could have been recreated in the meantime. Just open() it, the kernel will tell you whether it exists or not.

There are more points that currently slip my mind, hope this explains it somewhat.

Reply:
In addition to Timo's comments, `link.AttachRawLink` is implicitly feature probing for bpf_link here: 
https://github.com/cilium/ebpf/blob/b17ecdb6d25df55a953c156f66f5d59fb44a7d03/link/link.go#L178-L181
That is the piece of code that would return `ErrNotSupported` on kernels < 5.7.

On kernels >= 5.7 I agree with the 'ask for forgiveness' approach to see if the agent has previously attached progs with `PROG_ATTACH`